File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        mincost = float('inf')
        maxprofit = 0
        for p in prices:
            mincost = min(mincost, p)
            maxprofit = max(maxprofit, p - mincost)
        return maxprofit

File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        // int s0 = 0; // start state profit, always 0
        int s1 = INT_MIN; // buy state profit
        int ans = 0; // sell state profit
        for (int i: prices) {
            s1 = max(s1, -i);
            ans = max(ans, s1 + i);
        }
        return ans;
    }
};